libc/src/__support/macros/sanitizer.h

```diff
@@ -47,14 +47,21 @@
 // Functions to unpoison memory
 //-----------------------------------------------------------------------------
 
-#if defined(LIBC_HAVE_MEMORY_SANITIZER) && __has_builtin(__builtin_constant_p)
-// Only perform MSAN unpoison in non-constexpr context.
+#if defined(LIBC_HAVE_MEMORY_SANITIZER)
+// Only perform MSAN unpoison in non-constexpr context and silence
+// '-Wconstant-evaluated' when MSAN_UNPOISON is called from manifestly constant
+// contexts.
 #include <sanitizer/msan_interface.h>
 #define MSAN_UNPOISON(addr, size)                                              \
   do {                                                                         \
-    if (!__builtin_constant_p(*addr)) {                                        \
+    _Pragma("GCC diagnostic push \"-Wconstant-evaluated\"");                   \
+    _Pragma("GCC diagnostic ignored \"-Wconstant-evaluated\"");                \
+    _Pragma("clang diagnostic push \"-Wconstant-evaluated\"");                 \
+    _Pragma("clang diagnostic ignored \"-Wconstant-evaluated\"");              \
```

Comment:
```
<source>:4:20: warning: '__builtin_is_constant_evaluated' will always evaluate to 'true' in a manifestly constant-evaluated expression [-Wconstant-evaluated]
    4 |     if constexpr (!__builtin_is_constant_evaluated())
      |                    ^
```
https://godbolt.org/z/c513d1oPq

I'm not sure this is the correct pattern here, but don't yet have enough experience with constexpr to know how to handle what you're trying to do. Let me think about this, or ask more people for advice.

Reply:
I think if you simply remove the `constexpr` from the `if constexpr` (L61), then you will no longer observe the warning, making the pragmas unnecessary.

https://godbolt.org/z/jfPzEonKd